Wscc 112 troubleshooting guide (#1394)
mulesoft/docs-connectors: web-service-consumer/1.6/modules/ROOT/pages/web-service-consumer-troubleshooting.adoc

= Troubleshoot Web Service Consumer Connector - Mule 4

> To troubleshoot Anypoint Connector for Web Service Consumer (Web Service Consumer Connector), review the following sections for performing general troubleshooting, addressing specific common errors, and interpreting commonly thrown messages.

Reviewer: Per the style ref., we want to avoid "self reference": things like "this guide," "the following <separate> section/topic/example/etc.," and so on
 ```suggestion
To troubleshoot Anypoint Connector for Web Service Consumer (Web Service Consumer Connector), become familiar with the information about performing general troubleshooting, addressing specific common errors, and interpreting commonly thrown messages.
```
I apologize if I missed this during my template review!

Reviewer: HOWEVER...THIS topic has four specific subtopics, so why not name them: "...about enabling wire logging, troubleshooting access attachments, troubleshooting design-time WSDL issues, and interpreting commonly thrown messages?
Also...should the first topic really be titled Troubleshoot Transport Requests and Responses, to be parallel with its peers?

== Enable Wire Logging

> When troubleshooting Web Service Consumer Connector it is helpful to look at the transport request and response on the Web Service Consumer.

Reviewer: This is another of those modification errors: "it" is not doing the troubleshooting. More precisely:
```suggestion
To begin troubleshooting Web Service Consumer Connector, look at the transport request and response on the Web Service Consumer.
```

> To begin troubleshooting Web Service Consumer Connector, look at the transport request and response on the Web Service Consumer.

Reviewer 2: "transport,"

> If you are using the underlying HTTP transport, you can enable wire logging to see the exact HTTP message, the headers used in the transmission and the status of the response.

Reviewer: ```suggestion
If you are using the underlying HTTP transport, you can enable wire logging to see the exact HTTP message, the transmission headers, and the response status.
```

To enable wire logging:

> . Access Anypoint Studio, navigate to the *Package Explorer* view.

Reviewer: ```suggestion
. Access Anypoint Studio and navigate to the *Package Explorer* view.
```

. Open your application's project name.

> . Open the *src/main/resources* path folder.

Reviewer: I am used to seeing file names and paths and directories in monospace font, as "syntaxy" things.

> . Open the *src/main/resources* path folder.

Reviewer 2: For cloud connector doc, we use backticks for paths and file names.

. Open the `log4j2.xml` file inside the folder.

> . Add or uncomment the following line:

Reviewer: By this do we mean that if this line is already there, uncomment it to enable it; otherwise, add it?
I was slowed down a bit by "add or uncomment"...but maybe that is just me?

+
[source,xml,linenums]
----
<AsyncLogger name="org.mule.service.http.impl.service.HttpMessageLogger" level="DEBUG" />
----
+

> . Save all your application changes.

Reviewer: ```suggestion
. Save your changes.
```

> .  Click the project name in *Package Explorer* and then click *Run* > *Run As* > *Mule Application*.

Reviewer: ```suggestion
. Click the project name in *Package Explorer* and then click *Run* > *Run As* > *Mule Application*.
```

> . Navigate to the *Console* view to read the HTTP messages in the Mule's logger, for example:

Reviewer: ```suggestion
. Navigate to the *Console* view to read the HTTP messages in the logger:
```

User-Agent: AHC/1.0
Connection: keep-alive
Accept: */*
Content-Type: text/xml; charset=UTF-8
Transfer-Encoding: chunked

b1
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"><soap:Body><ns0:listAllFlights xmlns:ns0="http://soap.training.mulesoft.com/"/></soap:Body></soap:Envelope>

DEBUG 2021-04-06 15:04:23,058 [[wsc-base].wsc-dispatcher.11 SelectorRunner] [processor: ; event: ] org.mule.service.http.impl.service.HttpMessageLogger.wsc-dispatcher: REQUESTER
0

DEBUG 2021-04-06 15:04:23,319 [[wsc-base].wsc-dispatcher.11 SelectorRunner] [processor: ; event: ] org.mule.service.http.impl.service.HttpMessageLogger.wsc-dispatcher: REQUESTER
HTTP/1.1 200
Date: Tue, 06 Apr 2021 18:04:23 GMT
Content-Type: text/xml; charset=UTF-8
Transfer-Encoding: chunked
Connection: keep-alive
MULE_ENCODING: UTF-8
Strict-Transport-Security: max-age=31536000; includeSubdomains;
X-Content-Type-Options: 1
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block

bfe
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"><soap:Body><ns2:listAllFlightsResponse xmlns:ns2="http://soap.training.mulesoft.com/">...
----

== Troubleshoot Access Attachments

> Some servers can respond with MTOM messages. These messages separate their data into parts which enables different content types of data to coexist in the same HTTP message.

Reviewer: Respond to requests?
```suggestion
Some servers respond to requests by sending MTOM messages. These messages separate their data into parts, which enables different types of content data to coexist in the same HTTP message.
```

Reviewer: Should we spell out the first instance of MTOM?

> MTOM is in charge of sending binary data in a separate part and referencing the part from the corresponding field within the SOAP envelope.

Reviewer: ```suggestion
MTOM sends binary data separately from other data, and references that part in the corresponding field within the SOAP envelope.
```

Reviewer: Please doublecheck all suggestions in this MTOM area.

> Web Service Consumer users interact with these messages (multipart) only with the MTOM configuration. This is not to say that MTOM, XOP, and MIME or multipart are the same.

Reviewer: Can you provide some transitions in lines 64-67? The text goes from introducing MTOM as able to combine diff. types of info by using parts. Then it talks about a separate binary data part, and SOAP. Then it returns to the multipart message and Web Service (so why the separate, binary part). What does "only with the MTOM configuration" mean: what part of all this uses the MTOM configuration, and what do we mean by that configuration (binary part, multiple parts)? Finally, why go off topic to mention XOP, MIME, and "multipart" (and why is "multipart" separate from MTOM)?

Reviewer: If I had to guess, I'd say that you took this info from the sources without reworking it your Kevo way!

> Web Service Consumer users interact with these messages (multipart) only with the MTOM configuration. This is not to say that MTOM, XOP, and MIME or multipart are the same.

Reviewer 2: Do you need to say that these things are not the same? I am not sure why you need to say this.

Maybe "Web Service Consumer users use the MTOM configuration to interact with these messages. However, MTOM, XOP, and MIME or multipart are not the same.

Is multipart part of MIME? If not "...MIME, and multipart are not the same."

Refer to the following documentation for further information:

* https://www.ietf.org/rfc/rfc2387.txt[RFC2387 - The MIME Multipart/Related Content-type]
* https://www.w3.org/TR/2005/REC-xop10-20050125/[XML-binary Optimized Packaging (XOP) W3C Recommendation]
* https://www.w3.org/TR/soap12-mtom/[SOAP Message Transmission Optimization Mechanism (MTOM) W3C Recommendation]

> Since version 1.6.4, Web Service Consumer Connector enables you to configure MTOM for request messages, that also automatically manages MTOM responses regardless of the MTOM setting in the Web Service Consumer global configuration. Thereby, incoming messages with MTOM are treated as MTOM enabled regardless of whether MTOM is configured in the Web Service Consumer global configuration.

Reviewer: GLOBALLY avoid "since" because of its multiple meanings. Also, doesn't v1.6.4 refer to the Consumer and not the Connector?
```suggestion
Beginning with version 1.6.4, Web Service Consumer Connector enables you to configure MTOM for request messages that also automatically manage MTOM responses, regardless of the MTOM setting in the Web Service Consumer global configuration. Incoming messages with MTOM are treated as MTOM-enabled regardless of whether MTOM is configured in the Web Service Consumer global configuration.
```
If this is so...why do we need all the previous text about having to configure MTOM in WSC?
I think you want to revisit this text section, avoiding all the MTOM background and saying something like before 1.6.4 of WSC, you had to enable MTOM in WSC config, but now you....

> Beginning with version 1.6.4, Web Service Consumer Connector enables you to configure MTOM for request messages that also automatically manage MTOM responses, regardless of the MTOM setting in the Web Service Consumer global configuration. Incoming messages with MTOM are treated as MTOM-enabled regardless of whether MTOM is configured in the Web Service Consumer global configuration.

Reviewer 2: enabled, regardless

The following example shows the Web Service Consumer Connector global configuration with MTOM enabled:

.Web Service Consumer Connector Global Element
image::wsc-troubleshoot-3.png[Web Service Consumer global element configuration with MTOM field set to TRUE]

In the XML editor, the `<wsc:config>` and `mtomEnabled` configurations look like this:

[source,xml,linenums]
----
<wsc:config name="Web_Service_Consumer_Config" doc:name="Web Service Consumer Config">
    <wsc:connection wsdlLocation="http://ilt.mulesoft-training.com/essentials/delta?wsdl" service="TicketServiceService" port="TicketServicePort" address="http://ilt.mulesoft-training.com/essentials/delta" mtomEnabled="true">
        <wsc:web-service-security actor="http://schemas.xmlsoap.org/soap/actor/next" />
    </wsc:connection>
</wsc:config>
----

> Depending on the server side multipart setting, the server may decide to send some binary data as a base64-encoded text field or as a referenced binary part in the MIME message.

Reviewer: ```suggestion
Depending on its multipart setting, the server might send some binary data as a Base64-encoded text field or as a referenced binary part in the MIME message.
```

To troubleshoot how to access the binary data:

> * If the server sends the binary data as base64-encoded text field, you must access this field as part of the message. +

Reviewer: ```suggestion
* If the server sends the binary data as a Base64-encoded text field, you must access this field as part of the message. +
```

> Base64 decoding is not required because the Web Service Consumer performs the decoding by itself.

Reviewer: ```suggestion
Base64 decoding is automatically performed by Web Service Consumer.
```

* If the server sends the binary data as a referenced binary part, you must access this field as an attachment.
+

> The following example shows an image that was sent as an attachment with MTOM by a Web service server. You use DataWeave to access and transform the image into JSON. In this case, Base64 encoding is required:

Reviewer: ```suggestion
The following example shows an image sent as an attachment with MTOM by a web service server. You use DataWeave to access and transform the image into JSON. In this case, Base64 encoding is required:
```

+
[source,DataWeave,linenums]
----
import * from dw::core::Strings
import dw::core::Binaries
output application/json
ns nam http://mynamespace.com
---
{
...
"image": Binaries::toBase64(payload.attachments[0]),
...
}
----

== Troubleshoot Design Time WSDL Issues

> Some errors related to incorrect WSDL recognition, acquisition, or parsing may lead to some of the following issues:

Reviewer: ```suggestion
Errors related to incorrect WSDL recognition, acquisition, or parsing sometimes lead to the following issues:
```

* DataSense does not correctly expose the message metadata defined in the WSDL.

> * In the Web Service Consumer Connector configuration screen, the *Operations* field does not show any Web service operation names that the connector aims to invoke:

Reviewer: ```suggestion
* In the Web Service Consumer Connector configuration screen, the *Operations* field does not show any web service operation names for the connector to invoke:
```

+
.Web Service Consumer Connector Operations field
image::wsc-troubleshoot-1.png[Web Service Consumer Connector Operations field shows empty]

* In the Web Service Consumer Connector global element configuration, the *Port* and *Address* fields are not automatically filled when a WSDL location is provided.
+
.Web Service Consumer Connector Global Element
image::wsc-troubleshoot-2.png[In the Web Service Consumer Connector global element window, the port and address fields are not filled when a WSDL location is selected]

To resolve these issues:

* Check the integrity of your WSDL by using your preferred online or desktop viewer applications. +

> Every parsing error during the configuration it not shown at design time, so an external application is needed.

Reviewer: ```suggestion
Because every parsing error during the configuration is not shown during design, implementing an external application helps detect them.
```

> * If you do not add the WSDL as a resource in the application's *src/main/resources* folder in Studio, check for a correct connection to the server hosting the WSDL. +

Reviewer: ditto question about why the path name is not in mono

Reviewer: Does "check for a correct connection to the server hosting the WSDL" mean "check to verify the connection to the server hosting the WSDL"?

> * If you do not add the WSDL as a resource in the application's *src/main/resources* folder in Studio, check for a correct connection to the server hosting the WSDL. +

Reviewer 2: The path should be in backticks.

> Sometimes you need to add custom HTTP settings to access an HTTP server in an HTTP security layer, such as OAuth settings or user/password settings. See xref:web-service-consumer-configure.adoc[Setting a Custom HTTP Transport Configuration.]

Reviewer: ```suggestion
Sometimes you need to add custom HTTP settings, such as OAuth settings or user/password settings, to access an HTTP server in an HTTP security layer. See xref:web-service-consumer-configure.adoc[Setting a Custom HTTP Transport Configuration.]
```

[[common-throws]]
== Understand Common Throws

> Here is a list of common throws messages and how to interpret them.

Reviewer 2: Use "throw messages", "throws," or "messages" instead of "throws messages."

* WSC:SOAP_FAULT

  Error matching the SOAP response with the format provided by the WSDL.

>   Every CXF SOAP fault error is wrapped in a WSC:SOAP_FAULT.

Reviewer: Is there a reason to separate these sentences? Ditto below.

Reviewer: ```suggestion
An error occurred during matching the SOAP response with the format provided by the WSDL. Every CXF SOAP fault error is wrapped in a WSC:SOAP_FAULT.
```
"wrapped in a" is a little idiomatic; can you think of an alternative?

>   Every CXF SOAP fault error is wrapped in a WSC:SOAP_FAULT.

Reviewer 2: Maybe backtick WSC:SOAP_FAULT? Is that a throw type? It would be good to say what it is.

* WSC:BAD_REQUST

>   The Web Service Consumer Connector operation does not exist in the WSDL.
> 
>   The request body is not a valid XML.

Reviewer: ```suggestion
The Web Service Consumer Connector operation does not exist in the WSDL. The request body uses invalid XML.
```

* WSC:INVALID_WSDL

>   Bad formatted WSDL.

Reviewer: ```suggestion
The WSDL is poorly formatted.
```

>   The request body is not a valid XML.

Reviewer 2: I don't see Kim's change here.

Reviewer 3: That's because this error can have two meanings, thus I added each meaning in different lines.
